asn1tools/compiler.py

```diff
@@ -167,6 +167,42 @@ def decode(self, name, data, check_constraints=False):
 
         return decoded
 
+    def decode_with_length(self, name, data, check_constraints=False):
+        """Decode given bytes object `data` as given type `name` and return
+        the decoded data as a dictionary, along with byte length of the data.
+
+        Use to get length of indefinite-length BER encoded data.
+        Only works for BER or DER codec
+
+        If `check_constraints` is ``True`` all objects in `data` are
+        checked against their ASN.1 type constraints. A
+        ConstraintsError exception is raised if the constraints are
+        not fulfilled. Set `check_constraints` to ``False`` to skip
+        the constraints check and minimize the runtime overhead, but
+        instead allow decoding of values not fulfilling the
+        constraints.
+
+        >>> foo.decode_with_length('Question', b'0\\x0e\\x02\\x01\\x01\\x16\\x09Is 1+1=3?')
+        ({'id': 1, 'question': 'Is 1+1=3?'}, 16)
+
+        """
+
+        try:
+            type_ = self._types[name]
+        except KeyError:
+            raise DecodeError(
+                "Type '{}' not found in types dictionary.".format(name))
+
+        if not hasattr(type_, 'decode_with_length'):
+            raise DecodeError("Type '{}' does not support decode_with_length".format(name))
```

Comment:
Make all types have this method. Those that does not implement it should return ``NotImplementedError(a message)``

Reply:
I was going to add abstract decode_with_length() to base Type class but it doesn't even have abstract .decode() method?
And will a top-level type of BER Specification always be a CompiledType (wrapping another type)?

Reply:
Maybe I forgot to add the decode method. It's a mistake in that case.

I think so, yes. It was quite some time since I worked on this project, so I might be wrong.

Reply:
Base Type for all codecs do not have abstract .decode() method.
Also CompileType behaves differently to standard Type (acts as entry point for decoding?), and its .encode() bahaves differently (just calls wrapped type .decode() and strips the returned offset)
So I don't think it makes sense to add decode_with_length() to all Types (Types already return data and offset)
I can add abstract decode_with_length() to base CompiledType class, however it also does not have an abstract .decode()

Reply:
Yeah it should be added to the wrapper class CompiledType. Try to add it and implement a test calling it for a codec that does not implement it. If the expected exception is raised we are ok.

Reply:
Looks like all codec implementations of CompiledType share a common pattern (setting self._type on __init__, type() property, __repr__ implementation), looks like this should be moved to base CompiledType class? along with abstract .decode() and .decode_with_length()

Reply:
Yep done